tests/functional/container/test_replication.py

```diff
@@ -0,0 +1,155 @@
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library.
+
+import os
+import time
+from oio.api.object_storage import ObjectStorageApi
+from tests.utils import BaseTestCase, random_str
+
+
+class TestContainerReplication(BaseTestCase):
+    """
+    Test container replication, especially what happens when one copy
+    has missed some operations (service has been down and is back up),
+    or has been lost (database file deleted).
+    """
+
+    down_cache_opts = {'client.down_cache.avoid': 'false',
+                       'client.down_cache.shorten': 'true'}
+
+    def setUp(self):
+        super(TestContainerReplication, self).setUp()
+        if int(self.conf.get('container_replicas', 1)) < 3:
+            self.skipTest('Container replication must be enabled')
+        self.api = ObjectStorageApi(self.ns, pool_manager=self.http_pool)
+        self.must_restart_meta2 = False
+        self.wait_for_score(('meta2', ))
+        self._apply_conf_on_all('meta2', self.__class__.down_cache_opts)
+
+    @classmethod
+    def tearDownClass(cls):
+        # Be kind with the next test suites
+        cls._cls_reload_proxy()
+        time.sleep(3)
+        cls._cls_reload_meta()
+        time.sleep(1)
+
+    def tearDown(self):
+        # Start all services
+        self._service('@' + self.ns, 'start')
+        super(TestContainerReplication, self).tearDown()
+        # Restart meta2 after configuration has been reset by parent tearDown
+        if self.must_restart_meta2:
+            self._service('@meta2', 'restart', wait=2.0)
+
+    def _apply_conf_on_all(self, type_, conf):
+        all_svc = [x['addr'] for x in self.conf['services'][type_]]
+        for svc in all_svc:
+            self.admin.service_set_live_config(svc, conf)
+
+    def _synchronous_restore_allowed(self):
+        dump_max_size = int(self.ns_conf.get('sqliterepo.dump.max_size',
+                                             1073741824))
+        return dump_max_size
+
+    def _test_restore_after_missed_diff(self):
+        cname = 'test_restore_' + random_str(8)
+        # Create a container
+        self.api.container_create(self.account, cname)
+        # Locate the peers
+        peers = self.api.directory.list(self.account, cname,
+                                        service_type='meta2')
+        # Stop one peer
+        kept = peers['srv'][0]['host']
+        stopped = peers['srv'][1]['host']
+        self.api.logger.info('Stopping meta2 %s', stopped)
+        self._service(self.service_to_gridinit_key(stopped, 'meta2'), 'stop')
+        # Create an object
+        self.api.object_create_ext(self.account, cname,
+                                   obj_name=cname, data=cname)
+        # Start the stopped peer
+        self.api.logger.info('Starting meta2 %s', stopped)
+        # FIXME(FVE): review all the delays
+        self._service(self.service_to_gridinit_key(stopped, 'meta2'),
```

Comment:
I suggest to wait here for the service to get a positive score, instead of relying on a "sleep()" in the `_service` method. It might take longer but it would give more certainty on the service responsiveness.

Reply:
I will do that.